cloud/connecting-algorithm/src/app.py

```diff
@@ -22,23 +55,96 @@ def lambda_handler(event, context):
                 }
             }
         )
-        print(allUsersQuery.get('Items'))
+        users = allUsersQuery.get('Items')
 
-        testUserId = '17d2f33d-67e0-40ab-977e-73d1580e990d'
+        # Remove 'Joker' user if there are an odd number of users to be matched
+        if len(users) % 2 == 1:
+            users = [user for user in users if user['primaryKey']['S'] != ('USER#' + JOKER_USER_ID)]
 
-        # Get user from dynamo
-        response = client.get_item(
-            TableName='TuesHey',
-            Key={
-                'primaryKey': {
-                    'S': 'USER#' + testUserId
-                },
-                'sortKey': {
-                    'S': 'METADATA#' + testUserId
+        prev_matches = []
+        for user in users:
+            # Generate date of which last match will be considered
+            today = datetime.date.today()
+            match_limit = today - datetime.timedelta(days=7*MATCH_COOLDOWN)
+            iso_date_match_limit = match_limit.isoformat()
+
+            # Get all matches of specific user after match_limit
+            matchHistoryQuery = client.query(
+                TableName='TuesHey',
+                KeyConditionExpression='primaryKey = :primarykeyval AND '\
+                                        'sortKey BETWEEN :sortkeyval1 AND :sortkeyval2',
+                ExpressionAttributeValues={
+                    ':primarykeyval': {
+                        'S': user['primaryKey']['S']
+                    },
+                    ':sortkeyval1': {
+                        'S': 'MATCH#' + iso_date_match_limit 
+                    },
+                    ':sortkeyval2': {
+                        'S': 'METADATA'
+                    }
                 }
-            }
-        )
-        print(response.get('Item'))
+            )   
+            prev_matches_for_user = matchHistoryQuery.get('Items')
+
+            # Sort previous matches in descending order according to date of match
+            prev_matches_for_user.sort(key=lambda match: match['sortKey']['S'], reverse=True)
+
+            # Append to prev_matches a list of the IDs of the users the current user was matched with
+            prev_matches.append([match['user2Id']['S'] for match in prev_matches_for_user])
```

Comment:
Line 94 is just basically stripping each MATCH object in `prev_matches_for_user` to just the `user2Id` property and storing a list of that to `prev_matches`, just to make it a bit easier in the function `eval_cost`.

So for example, if I were to print out `prev_matches`, I would get something like this:
`[['USER#c8345096-448d-4939-8691-c62259470667', 'USER#83afa77d-4209-49bb-853c-6010f555c6dd', 'USER#a6a4588c-224f-400a-b7eb-ac64b32732dc', 'USER#1f44e322-8fe3-4357-bf0f-aa0a82e120a7', 'USER#aac4759e-5e40-4479-a9ea-0bdc0647e655'], ['USER#83afa77d-4209-49bb-853c-6010f555c6dd', ...]]`

I could remove this, and just append `prev_matches_for_user` straight away to `prev_matches`.

Reply:
That makes sense. I want to avoid doing another loop over each match per user here. Just before `cost += constants.COST_PREV_MATCHED` in eval_cost, can we deconstruct it to just user2Id?